File: web_server/reference/web_server.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <filesystem>

namespace fs = std::filesystem;

const int BUFFER_SIZE = 1024;
const int PORT = 6789;

// 发送文件内容
void send_file(int socket, const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file.is_open()) {
        throw std::runtime_error("File not found");
    }

    // 读取文件内容
    std::string content((std::istreambuf_iterator<char>(file)), 
                        std::istreambuf_iterator<char>());

    // 构建HTTP响应
    std::string response = 
        "HTTP/1.1 200 OK\r\n"
        "Content-Type: text/html\r\n"
        "Content-Length: " + std::to_string(content.size()) + "\r\n"
        "\r\n" + content;

    if (send(socket, response.c_str(), response.size(), 0) < 0) {
        throw std::runtime_error("Failed to send file");
    }
}

// 发送404错误
void send_404(int socket) {
    std::string response = 
        "HTTP/1.1 404 Not Found\r\n"
        "Content-Type: text/html\r\n"
        "\r\n"
        "<html><body><h1>404 Not Found</h1></body></html>";

    if (send(socket, response.c_str(), response.size(), 0) < 0) {
        std::cerr << "Failed to send 404 response" << std::endl;
    }
}

// 解析HTTP请求获取文件名
std::string parse_filename(const char* request) {
    std::string req_str(request);
    size_t start = req_str.find(' ') + 1;
    size_t end = req_str.find(' ', start);
    std::string path = req_str.substr(start, end - start);
    
    // 移除路径开头的'/'
    if (!path.empty() && path[0] == '/') {
        path = path.substr(1);
    }
    
    // 默认返回index.html
    if (path.empty()) {
        path = "index.html";
    }
    
    return path;
}

int main() {
    // 创建socket
    int server_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket < 0) {
        std::cerr << "Socket creation failed" << std::endl;
        return 1;
    }

    // 设置socket选项
    int opt = 1;
    if (setsockopt(server_socket, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
        std::cerr << "Setsockopt failed" << std::endl;
        return 1;
    }

    // 绑定地址和端口
    sockaddr_in server_addr{};
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = INADDR_ANY;
    server_addr.sin_port = htons(PORT);

    if (bind(server_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        std::cerr << "Bind failed" << std::endl;
        return 1;
    }

    // 开始监听
    if (listen(server_socket, 1) < 0) {
        std::cerr << "Listen failed" << std::endl;
        return 1;
    }

    std::cout << "Server ready on port " << PORT << std::endl;

    while (true) {
        std::cout << "Ready to serve..." << std::endl;

        // 接受客户端连接
        sockaddr_in client_addr{};
        socklen_t client_len = sizeof(client_addr);
        int client_socket = accept(server_socket, (struct sockaddr*)&client_addr, &client_len);
        
        if (client_socket < 0) {
            std::cerr << "Accept failed" << std::endl;
            continue;
        }

        // 接收HTTP请求
        char buffer[BUFFER_SIZE] = {0};
        ssize_t bytes_read = recv(client_socket, buffer, BUFFER_SIZE, 0);
        
        if (bytes_read < 0) {
            std::cerr << "Receive failed" << std::endl;
            close(client_socket);
            continue;
        }

        try {
            // 解析文件名
            std::string filename = parse_filename(buffer);
            
            // 发送文件
            send_file(client_socket, filename);
        } catch (const std::exception& e) {
            std::cerr << "Error: " << e.what() << std::endl;
            send_404(client_socket);
        }

        close(client_socket);
    }

    close(server_socket);
    return 0;
}

File: web_server/reference/web_server.py
#import socket module
from socket import *
import sys # In order to terminate the program

serverSocket = socket(AF_INET, SOCK_STREAM)

# Prepare a sever socket
# Fill in start
serverPort = 6789  # You can change this port if needed
serverSocket.bind(('', serverPort))  # Bind the server socket to all interfaces and port
serverSocket.listen(1)  # Listen for incoming connections (max 1 connection at a time)
# Fill in end

while True:
    # Establish the connection
    print('Ready to serve...')
    connectionSocket, addr = serverSocket.accept()  # Fill in start -> accept incoming connection
    # Fill in end
    
    try:
        breakpoint()
        message = connectionSocket.recv(1024).decode()  # Fill in start -> receive HTTP request
        # Fill in end
        
        filename = message.split()[1]
        f = open(filename[1:])  # Remove leading '/'
        outputdata = f.read()  # Fill in start -> read requested file
        # Fill in end

        # Send one HTTP header line into socket
        # Fill in start
        response_header = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n"
        connectionSocket.send(response_header.encode())
        # Fill in end

        # Send the content of the requested file to the client
        for i in range(0, len(outputdata)):
            connectionSocket.send(outputdata[i].encode())
        connectionSocket.send("\r\n".encode())

        connectionSocket.close()

    except IOError:
        # Send response message for file not found
        # Fill in start
        response_404 = "HTTP/1.1 404 Not Found\r\nContent-Type: text/html\r\n\r\n<html><body><h1>404 Not Found</h1></body></html>"
        connectionSocket.send(response_404.encode())
        # Fill in end

        # Close client socket
        # Fill in start
        connectionSocket.close()
        # Fill in end

serverSocket.close()
sys.exit()  # Terminate the program after sending the corresponding data